terraform/main.tf

```diff
@@ -0,0 +1,117 @@
+provider "aws" {
+  region = "us-east-1"
+  profile = "serverless"
+}
+
+data "aws_vpc" "default" {
+  vpc_id = "vpc-0d60ef77"
+}
+
+data "aws_subnet_ids" "all" {
+  vpc_id = data.aws_vpc.default.id
+}
+
+resource "aws_elasticache_cluster" "network-data" {
+  cluster_id           = "network-data"
+  engine               = "redis"
+  node_type            = "cache.t3.small"
+  num_cache_nodes      = 1
+  parameter_group_name = "default.redis6.x"
+  engine_version       = "6.x"
+  port                 = 6379
+}
+
+resource "aws_ecr_repository" "data-collector" {
+  name                 = "data-collector"
+  image_tag_mutability = "MUTABLE"
+}
+
+resource "aws_iam_role" "ec2_role_data_collector" {
+  name = "ec2_role_data_collector"
+
+  assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "ec2.amazonaws.com"
+      },
+      "Effect": "Allow"
+    }
+  ]
+}
+EOF
+
+  tags = {
+    project = "hello-world"
+  }
+}
+
+resource "aws_iam_instance_profile" "ec2_profile_data_collector" {
+  name = "ec2_profile_data_collector"
+  role = aws_iam_role.ec2_role_data_collector.name
+}
+
+resource "aws_iam_role_policy" "ec2_policy" {
+  name = "ec2_policy"
+  role = aws_iam_role.ec2_role_data_collector.id
+
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": [
+        "ecr:GetAuthorizationToken",
+        "ecr:BatchGetImage",
+        "ecr:GetDownloadUrlForLayer"
+      ],
+      "Effect": "Allow",
+      "Resource": "*"
+    }
+  ]
+}
+EOF
+}
+
+module "dev_ssh_sg" {
+  source = "terraform-aws-modules/security-group/aws"
+
+  name        = "ec2_sg"
+  description = "Security group for ec2_sg"
+  vpc_id      = data.aws_vpc.default.id
+
+  ingress_cidr_blocks = ["127.0.0.1/32"]
+  ingress_rules       = ["ssh-tcp"]
+}
+
+data "aws_ami" "amazon_linux_2" {
+  most_recent = true
+  owners      = ["amazon"]
+
+  filter {
+    name   = "name"
+    values = ["amzn2-ami-hvm-*-x86_64-ebs"]
+  }
+}
+
+resource "aws_instance" "web" {
+  ami           = data.aws_ami.amazon_linux_2.id
+  instance_type = "t3.small"
+
+  root_block_device {
+    volume_size = 8
+  }
+
+  vpc_security_group_ids = [
+    module.dev_ssh_sg.this_security_group_id
+  ]
+  iam_instance_profile = aws_iam_instance_profile.ec2_profile.name
+
+  key_name                = "data-collector-key"
+  monitoring              = true
+  disable_api_termination = false
+  ebs_optimized           = true
```

Comment:
Do this option generates additional costs?
If so, why do we need an EBS optimized instance in this case?